File: haskell/src/Lexer.hs
module Lexer (
    tokenize
  , Token(..)
  , TokenType(..)
  )
  where

import Data.ByteString.Char8     (ByteString, cons, empty, (!?))
import Data.Char                 (isDigit, isLetter, isSpace)
import Data.Maybe                (fromMaybe)
import TokenTypes                (Lexer(..), Token(..), TokenType(..))
import Control.Lens              (makeLenses)
import Control.Lens.Operators    ((+~), (.~), (^.), (&))
import Control.Monad.Trans.State (State, modify, get, evalState)

makeLenses ''Lexer
makeLenses ''Token

tokenize :: ByteString -> [Token]
tokenize = evalState getTokens . (\x -> readChar $ Lexer x 0 0 '\0')

getTokens :: State Lexer [Token]
getTokens = do
  token <- getToken
  if   (token^.tokenType) == EOF
  then pure [token]
  else (token:) <$> getTokens

getToken :: State Lexer Token
getToken = do
  modify skipWhitespace
  lexer <- get
  let readToken = (modify readChar >>) . pure . Token
  case lexer^.ch of
    c | isLetter c || c == '_' ->  mapKeyword <$> readIdentifier
    c | isDigit c              ->  Token . INT <$> readDigit
    '='  -> readToken EQUAL
    '+'  -> readToken PLUS
    ','  -> readToken COMMA
    ';'  -> readToken SEMICOLON
    '('  -> readToken LPAREN
    ')'  -> readToken RPAREN
    '{'  -> readToken LBRACE
    '}'  -> readToken RBRACE
    '\0' -> readToken EOF
    _    -> readToken ILLEGAL

mapKeyword :: ByteString -> Token
mapKeyword "fn"     = Token FUNCTION
mapKeyword "let"    = Token LET
mapKeyword ident    = Token (IDENT ident)

readChar :: Lexer -> Lexer
readChar lexer =
  lexer & ch.~ '\0' `fromMaybe` ((lexer^.input) !? (lexer^.readPosition))
    & position.~(lexer^.readPosition)
    & readPosition+~1

readIdentifier :: State Lexer ByteString
readIdentifier = do
  lexer <- get
  if   isLetter (lexer^.ch) || lexer^.ch == '_'
  then modify readChar >> cons (lexer^.ch) <$> readIdentifier
  else pure empty

readDigit :: State Lexer ByteString
readDigit = do
  lexer <- get
  if   isDigit (lexer^.ch)
  then modify readChar >> cons (lexer^.ch) <$> readDigit
  else pure empty

skipWhitespace :: Lexer -> Lexer
skipWhitespace lexer
  | isSpace $ lexer^.ch = skipWhitespace $ readChar lexer
  | otherwise           = lexer

File: java/Lexer.java
import java.util.HashMap;
import java.util.Optional;

class Lexer {

    private final String input;
    private int pos = 0;

    final HashMap<String, Token> keywordMap = new HashMap<>() {
        {
            put("fn", createToken(TokenType.FUNC, "fn"));
            put("let", createToken(TokenType.LET, "let"));
        }
    };

    public Lexer(String input) {
        this.input = input;
    }

    public Token nextToken() {
        this.skipWhitespace();

        var tokenType = switch (this.getCc()) {
            case '{' -> TokenType.LSQIRLY;
            case '}' -> TokenType.RSQIRLY;
            case '(' -> TokenType.LPAREN;
            case ')' -> TokenType.RPAREN;
            case ',' -> TokenType.COMMA;
            case ';' -> TokenType.SEMI;
            case '+' -> TokenType.PLUS;
            case '=' -> TokenType.EQUAL;
            case '\0' -> TokenType.EOF;
            default -> TokenType.ILLEGAL;
        };

        var token = switch (tokenType) {
            case EOF -> this.createToken(tokenType, "eof");
            default -> this.createToken(tokenType, String.valueOf(this.getCc()));
        };

        if (Character.isLetter(this.getCc())) {
            var ident = this.indent();

            return Optional
                    .ofNullable(this.keywordMap.get(ident))
                    .orElseGet(() -> createToken(TokenType.IDENT, ident));
        } else if (Character.isDigit(this.getCc())) {
            return this.createToken(TokenType.INT, this.number());
        }

        this.advance();
        return token;
    }

    private void advance() {
        if (this.pos >= this.input.length()) {
            return;
        }
        this.pos++;
    }

    private char getCc() {
        if (this.pos >= this.input.length()) {
            return '\0';
        }
        return this.input.charAt(this.pos);
    }

    private Token createToken(TokenType type, String literal) {
        return new Token(type, literal);
    }

    private String number() {
        int pos = this.pos;
        while (Character.isDigit(this.getCc())) {
            this.advance();
        }
        return this.input.substring(pos, this.pos);
    }

    private String indent() {
        int pos = this.pos;
        while (Character.isLetter(this.getCc())) {
            this.advance();
        }
        return this.input.substring(pos, this.pos);
    }

    private void skipWhitespace() {
        while (Character.isWhitespace(this.getCc())) {
            this.advance();
        }
    }
}
